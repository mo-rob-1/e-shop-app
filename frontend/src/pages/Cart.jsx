import { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import CartItem from "../components/CartItem";
import { addToCart, removeFromCart } from "../redux/actions/cartActions";

function Cart() {
  const dispatch = useDispatch();

  const cart = useSelector((state) => state.cart);
  const { cartItems } = cart;

  useEffect(() => {}, []);

  const qtyChangeHandler = (id, qty) => {
    dispatch(addToCart(id, qty));
  };

  const removeFromCartHandler = (id) => {
    dispatch(removeFromCart(id));
  };

  const getCartCount = () => {
    return cartItems.reduce((qty, item) => Number(item.qty) + qty, 0);
  };

  const getCartSubTotal = () => {
    return cartItems.reduce((price, item) => price + item.price * item.qty, 0).toFixed(2);
  };
  return (
    <>
      <div className="checkout-container">
        <div>
          <h2 className="mb-10 font-bold text-2xl">Shopping Cart</h2>

          <div className="grid gap-5">
            {cartItems.length === 0 ? (
              <h2 className="text-xl font-bold">
                Your Cart Is Empty{" "}
                <Link className="underline" to="/">
                  Go Back
                </Link>
              </h2>
            ) : (
              cartItems.map((item) => (
                <CartItem
                  key={item.product}
                  item={item}
                  qtyChangeHandler={qtyChangeHandler}
                  removeHandler={removeFromCartHandler}
                />
              ))
            )}
          </div>
        </div>

        <div>
          <div className="shadow p-6 lg:p-10 rounded-xl mt-8">
            <p className="font-semibold mb-3">Subtotal ({getCartCount()}) items</p>
            <p className="font-bold text-2xl mb-3">Â£{getCartSubTotal()}</p>
            <div>
              <Link
                to="/checkout"
                className="bg-slate-800 text-white font-semibold inline-block pl-10 pr-10 pt-4 pb-4 mt-2 text-sm rounded-md"
              >
                Proceed To Checkout
              </Link>
            </div>
          </div>
        </div>
      </div>
    </>
  );
}

export default Cart;
