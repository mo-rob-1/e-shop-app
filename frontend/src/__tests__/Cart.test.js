import { render } from "@testing-library/react";
import Cart from "../pages/Cart";
import { createMemoryHistory } from "history";
import { Provider } from "react-redux";
import store from "../redux/store";
import { BrowserRouter as Router } from "react-router-dom";

describe("Cart", () => {
  it("renders the cart page", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Provider store={store} history={history}>
        <Router>
          <Cart />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders shopping cart title text", () => {
    const { getByText } = render(
      <Provider store={store}>
        <Router>
          <Cart />
        </Router>
      </Provider>
    );
    expect(getByText("Shopping Cart")).toBeInTheDocument();
  });

  it("renders cart empty text", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <Cart />
        </Router>
      </Provider>
    );

    const cartEmptyText = getByTestId("cart-empty-text");
    expect(cartEmptyText).toBeTruthy();
  });

  it("renders subtotal title text", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <Cart />
        </Router>
      </Provider>
    );
    expect(getByTestId("subtotal-title-text")).toBeInTheDocument();
  });

  it("renders subtotal price text", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <Cart />
        </Router>
      </Provider>
    );
    expect(getByTestId("subtotal-price-text")).toBeInTheDocument();
  });

  it("renders checkout button", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <Cart />
        </Router>
      </Provider>
    );
    expect(getByTestId("checkout-button")).toBeInTheDocument();
  });
});
