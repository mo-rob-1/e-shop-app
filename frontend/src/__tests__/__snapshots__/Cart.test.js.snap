// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart renders the cart page 1`] = `
<DocumentFragment>
  <div
    class="checkout-container"
  >
    <div>
      <h2
        class="mb-10 font-bold text-2xl"
      >
        Shopping Cart
      </h2>
      <div
        class="grid gap-5"
      >
        <h2
          class="text-xl font-bold"
          data-testid="cart-empty-text"
        >
          Your Cart Is Empty 
          <a
            class="underline"
            href="/"
          >
            Go Back
          </a>
        </h2>
      </div>
    </div>
    <div>
      <div
        class="shadow p-6 lg:p-10 rounded-xl mt-8"
      >
        <p
          class="font-semibold mb-3"
          data-testid="subtotal-title-text"
        >
          Subtotal (0) items
        </p>
        <p
          class="font-bold text-2xl mb-3"
          data-testid="subtotal-price-text"
        >
          Â£0.00
        </p>
        <div>
          <a
            class="bg-slate-800 text-white font-semibold inline-block pl-10 pr-10 pt-4 pb-4 mt-2 text-sm rounded-md"
            data-testid="checkout-button"
            href="/checkout"
          >
            Proceed To Checkout
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
